File: ts/sha1.ts
// Basic SHA-1 implementation in pure JavaScript.
function sha1(message: any) {
    function rotateLeft(n: any, s: any) {
        return (n << s) | (n >>> (32 - s));
    }

    function cvtHex(val: any) {
        let str = "";
        for (let i = 7; i >= 0; i--) {
            let v = (val >>> (i * 4)) & 0x0f;
            str += v.toString(16);
        }
        return str;
    }

    function utf8Encode(string: any) {
        return decodeURIComponent(encodeURIComponent(string));
    }

    let blockstart;
    let i, j;
    let W = new Array(80);
    let H0 = 0x67452301;
    let H1 = 0xefcdab89;
    let H2 = 0x98badcfe;
    let H3 = 0x10325476;
    let H4 = 0xc3d2e1f0;
    let A, B, C, D, E;
    let temp;

    message = utf8Encode(message);
    let msgLen = message.length;

    let wordArray = [];
    for (i = 0; i < msgLen - 3; i += 4) {
        j =
            (message.charCodeAt(i) << 24) |
            (message.charCodeAt(i + 1) << 16) |
            (message.charCodeAt(i + 2) << 8) |
            message.charCodeAt(i + 3);
        wordArray.push(j);
    }

    switch (msgLen % 4) {
        case 0:
            i = 0x080000000;
            break;
        case 1:
            i = (message.charCodeAt(msgLen - 1) << 24) | 0x0800000;
            break;
        case 2:
            i =
                (message.charCodeAt(msgLen - 2) << 24) |
                (message.charCodeAt(msgLen - 1) << 16) |
                0x08000;
            break;
        case 3:
            i =
                (message.charCodeAt(msgLen - 3) << 24) |
                (message.charCodeAt(msgLen - 2) << 16) |
                (message.charCodeAt(msgLen - 1) << 8) |
                0x80;
            break;
    }

    wordArray.push(i);
    while (wordArray.length % 16 != 14) wordArray.push(0);
    wordArray.push(msgLen >>> 29);
    wordArray.push((msgLen << 3) & 0x0ffffffff);

    for (blockstart = 0; blockstart < wordArray.length; blockstart += 16) {
        for (i = 0; i < 16; i++) W[i] = wordArray[blockstart + i];
        for (i = 16; i <= 79; i++)
            W[i] = rotateLeft(W[i - 3] ^ W[i - 8] ^ W[i - 14] ^ W[i - 16], 1);

        A = H0;
        B = H1;
        C = H2;
        D = H3;
        E = H4;

        for (i = 0; i <= 19; i++) {
            temp =
                (rotateLeft(A, 5) + ((B & C) | (~B & D)) + E + W[i] + 0x5a827999) &
                0x0ffffffff;
            E = D;
            D = C;
            C = rotateLeft(B, 30);
            B = A;
            A = temp;
        }

        for (i = 20; i <= 39; i++) {
            temp =
                (rotateLeft(A, 5) + (B ^ C ^ D) + E + W[i] + 0x6ed9eba1) & 0x0ffffffff;
            E = D;
            D = C;
            C = rotateLeft(B, 30);
            B = A;
            A = temp;
        }

        for (i = 40; i <= 59; i++) {
            temp =
                (rotateLeft(A, 5) +
                    ((B & C) | (B & D) | (C & D)) +
                    E +
                    W[i] +
                    0x8f1bbcdc) &
                0x0ffffffff;
            E = D;
            D = C;
            C = rotateLeft(B, 30);
            B = A;
            A = temp;
        }

        for (i = 60; i <= 79; i++) {
            temp =
                (rotateLeft(A, 5) + (B ^ C ^ D) + E + W[i] + 0xca62c1d6) & 0x0ffffffff;
            E = D;
            D = C;
            C = rotateLeft(B, 30);
            B = A;
            A = temp;
        }

        H0 = (H0 + A) & 0x0ffffffff;
        H1 = (H1 + B) & 0x0ffffffff;
        H2 = (H2 + C) & 0x0ffffffff;
        H3 = (H3 + D) & 0x0ffffffff;
        H4 = (H4 + E) & 0x0ffffffff;
    }

    let tempValue =
        cvtHex(H0) + cvtHex(H1) + cvtHex(H2) + cvtHex(H3) + cvtHex(H4);
    return tempValue.toLowerCase();
}

// HMAC function using SHA-1
function hmacSHA1(key: string, message: string) {
    const blockSize = 64; // Block size for SHA-1
    if (key.length > blockSize) {
        key = sha1(key); // Hash the key if it is longer than block size
    }
    if (key.length < blockSize) {
        key = key.padEnd(blockSize, String.fromCharCode(0)); // Pad key to block size
    }

    const oKeyPad = Array.from(key, (char: string) =>
        String.fromCharCode(char.charCodeAt(0) ^ 0x5c)
    ).join("");
    const iKeyPad = Array.from(key, (char: string) =>
        String.fromCharCode(char.charCodeAt(0) ^ 0x36)
    ).join("");

    return sha1(oKeyPad + sha1(iKeyPad + message));
}

// Function to generate a random hex string of a given length
function generateRandomHex(length: number) {
    const characters = "0123456789abcdef";
    let result = "";
    for (let i = 0; i < length; i++) {
        result += characters[Math.floor(Math.random() * characters.length)];
    }
    return result;
}

const length = 16;

export default function wrappedCryptoToken({
    salt = generateRandomHex(length),
    wrappedCryptoString = "",
}) {
    try {
        // Generate HMAC-SHA1 hash using our custom implementation
        const hash = hmacSHA1(salt, wrappedCryptoString);
        return {
            salt: salt,
            success: true,
            hash: Buffer.from(hash, "hex").toString("base64"),
        };
    } catch (err: any) {
        return {
            success: false,
            salt: null,
            hash: null,
            message: err.message,
        };
    }
}

File: ts/sha1.py
import hashlib
import random
import string

# Rotate left function
def rotate_left(n, b):
    return ((n << b) | (n >> (32 - b))) & 0xFFFFFFFF

# Function to convert a value to a hexadecimal string
def cvt_hex(val):
    return ''.join([f'{(val >> (i * 4)) & 0xF:x}' for i in range(7, -1, -1)])

# UTF-8 encoding
def utf8_encode(string):
    return string.encode('utf-8')

# SHA-1 hashing algorithm
def sha1(message):
    message = utf8_encode(message)
    msg_len = len(message)
    word_array = []

    # Prepare the word array
    for i in range(0, msg_len - 3, 4):
        j = (message[i] << 24) | (message[i+1] << 16) | (message[i+2] << 8) | message[i+3]
        word_array.append(j)

    i = {
        0: 0x080000000,
        1: (message[msg_len - 1] << 24) | 0x0800000,
        2: (message[msg_len - 2] << 24) | (message[msg_len - 1] << 16) | 0x08000,
        3: (message[msg_len - 3] << 24) | (message[msg_len - 2] << 16) | (message[msg_len - 1] << 8) | 0x80
    }.get(msg_len % 4, 0)

    word_array.append(i)
    
    while len(word_array) % 16 != 14:
        word_array.append(0)
    
    word_array.append((msg_len * 8) >> 32)
    word_array.append((msg_len * 8) & 0xFFFFFFFF)

    # Initialize variables
    H0 = 0x67452301
    H1 = 0xefcdab89
    H2 = 0x98badcfe
    H3 = 0x10325476
    H4 = 0xc3d2e1f0

    for blockstart in range(0, len(word_array), 16):
        W = word_array[blockstart:blockstart + 16] + [0] * 64
        for i in range(16, 80):
            W[i] = rotate_left(W[i - 3] ^ W[i - 8] ^ W[i - 14] ^ W[i - 16], 1)
        
        A = H0
        B = H1
        C = H2
        D = H3
        E = H4

        for i in range(80):
            if 0 <= i <= 19:
                f = (B & C) | (~B & D)
                k = 0x5A827999
            elif 20 <= i <= 39:
                f = B ^ C ^ D
                k = 0x6ED9EBA1
            elif 40 <= i <= 59:
                f = (B & C) | (B & D) | (C & D)
                k = 0x8F1BBCDC
            else:
                f = B ^ C ^ D
                k = 0xCA62C1D6

            temp = (rotate_left(A, 5) + f + E + k + W[i]) & 0xFFFFFFFF
            E = D
            D = C
            C = rotate_left(B, 30)
            B = A
            A = temp

        H0 = (H0 + A) & 0xFFFFFFFF
        H1 = (H1 + B) & 0xFFFFFFFF
        H2 = (H2 + C) & 0xFFFFFFFF
        H3 = (H3 + D) & 0xFFFFFFFF
        H4 = (H4 + E) & 0xFFFFFFFF

    return ''.join([cvt_hex(h) for h in [H0, H1, H2, H3, H4]]).lower()

# HMAC-SHA1 implementation
def hmac_sha1(key, message):
    block_size = 64
    if len(key) > block_size:
        key = sha1(key)
    if len(key) < block_size:
        key = key.ljust(block_size, '\0')

    o_key_pad = ''.join([chr(ord(c) ^ 0x5C) for c in key])
    i_key_pad = ''.join([chr(ord(c) ^ 0x36) for c in key])

    return sha1(o_key_pad + sha1(i_key_pad + message))

# Generate random hex string
def generate_random_hex(length):
    return ''.join(random.choice(string.hexdigits.lower()) for _ in range(length))

# Wrapped HMAC-SHA1 function
def wrapped_crypto_token(salt=None, wrapped_crypto_string=''):
    try:
        if salt is None:
            salt = generate_random_hex(16)

        hash_value = hmac_sha1(salt, wrapped_crypto_string)
        return {
            'salt': salt,
            'success': True,
            'hash': hashlib.sha1(hash_value.encode()).digest().hex()
        }
    except Exception as err:
        return {
            'success': False,
            'salt': None,
            'hash': None,
            'message': str(err)
        }

# Example usage
result = wrapped_crypto_token(salt='6f27a28e58f950b1',wrapped_crypto_string="ccccc")
x = result['salt']
print(x)
print(result)
